### PR TITLE
Added checked arithmetic extensions

### PR DESCRIPTION
**Pull request by L117**
This PR adds `checked_add` and `checked_sub` methods for `Date`, `PrimitiveDateTime`, and `OffsetDateTime`.
  
Part of #364

### CODE REVIEW

**Comment by jhpratt** [comment]
I presume you opted for a trait because you wanted the ability to pass either a `time::Duration` or `std::time::Duration`? I don't find it necessary to accept both. There is already the ability to convert between the two duration types. Aside from `NumericalStdDuration`, the only place stdlib types are (publicly) used in time is for conversions.

Do you have a use case for checked arithmetic on `Time`? That's really the only type that has an obvious behavior for overflow: wrapping (which is already done).

In short I'd prefer the following inherent implementations for checked arithmetic:

```rust
impl {Date, PrimitiveDateTime, OffsetDateTime} {
    pub const fn checked_add(self, rhs: Duration) -> Option<Self>;
    pub const fn checked_sub(self, rhs: Duration) -> Option<Self>;
}
```

The implementations for "regular" arithmetic could then be updated to call these methods internally.

**Comment by L117** [comment]
Not exactly. Personally I only ever use `time::Duration`, but `time` has implementations `Add<StdDuration>` and `Sub<StdDuration>`for {`Time`,`Date`,`PrimitiveDateTime`,`OffsetDateTime`} (No explicit conversion), so I followed this logic and added `Checked{Add,Sub}`. Also, I'd personally prefer to have trait impls (overloaded functions) over a bunch of long similar inherent methods, if any `time` structure will ever have checked_{add,sub} with multiple possibilities for argument.
Anyway, I can get rid of these traits without much regret (:

What about use case for checked arithmetic on `Time`, it is pretty obvious: `Time` +/- `Duration` is within the same day. Seems useful to me. Personally, I'd keep it just for consistency and document that time will always wrap otherwise.

So, what should I do?

**Comment by ahcodedthat** [comment]
How about a checked conversion from `SystemTime`? The implementation of `From<SystemTime>` for `OffsetDateTime` is also panicky.

**Comment by jhpratt** [comment]
@ahcodedthat This is the implementation

```rust
impl From<SystemTime> for OffsetDateTime {
    fn from(system_time: SystemTime) -> Self {
        match system_time.duration_since(SystemTime::UNIX_EPOCH) {
            Ok(duration) => Self::UNIX_EPOCH + duration,
            Err(err) => Self::UNIX_EPOCH - err.duration(),
        }
    }
}
```

I presume you're referring to the fact that the arithmetic can fail? If so, this would only happen if a user deliberately added a very large `Duration` to the `SystemTime` beforehand, as it's an opaque type.

File: tests/integration/offset_date_time.rs [hunk 1]
```diff
@@ -868,3 +868,203 @@ fn to_std() {
...
+    assert_eq!(
+        datetime!(2021 - 10 - 25 14:01:53.45 UTC).checked_add(-Duration::DAY * 2),
+        Some(datetime!(2021 - 10 - 23 14:01:53.45 UTC))
+    );
+    assert_eq!(
+        datetime!(2021 - 10 - 25 14:01:53.45 UTC).checked_add(-Duration::WEEK),
+        Some(datetime!(2021 - 10 - 18 14:01:53.45 UTC))
+    );
+
+    #[cfg(feature = "large-dates")]
```
[the remark this answers is not in the document]
Comment (jhpratt) [reply tests/integration/offset_date_time.rs:908]: I'll look into it, but as I mentioned I try to keep CI time low, both for my sanity and that of contributors. It's probably feasible to add a single run to the build matrix where it's tested with everything except `large-dates`.

File: tests/integration/date.rs [hunk 2]
```diff
@@ -767,3 +767,111 @@ fn regression_check() {
     assert_eq!(week, 1);
     assert_eq!(weekday, Weekday::Monday);
 }
+
+#[test]
+fn checked_add_duration() {
+    // Adding subday duration
+    assert_eq!(
+        Date::MIN.checked_add(Duration::new(84_399, 999_999_999)),
```
[the remark this answers is not in the document]
Comment (jhpratt) [reply tests/integration/date.rs]: Pushing a change up for this myself.

**Review by jhpratt — APPROVED** [review]
Looks good, just one minor change in addition to one previously mentioned. I've done this myself and will merge shortly (I've got some changes on `dev` that haven't been pushed to `main`).

File: tests/integration/date.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (L117) [reply tests/integration/date.rs]: Oh. Sorry. I was sure I have fixed this. Thanks.

**Comment by ahcodedthat** [comment]
> [@jhpratt] I presume you're referring to the fact that the arithmetic can fail? If so, this would only happen if a user deliberately added a very large `Duration` to the `SystemTime` beforehand, as it's an opaque type.

Right. A malicious actor could use this for a denial-of-service attack by intentionally overflowing the conversion.

* A local attacker might set a file's time stamp to something outside of `OffsetDateTime`'s range, then trigger a privileged Rust program to look at its time stamp and panic.
* A remote attacker might send a time stamp over the network to a Rust program that converts it first to `SystemTime` and then `OffsetDateTime`, again resulting in panic.

**Comment by jhpratt** [comment]
The standard library panics on a wide variety of items. The behavior is based on what is 99.999+% of the time true. There is no plausible reason someone would be adding an arbitrary, unchecked duration to a `SystemTime` that I can think of. If there is any evidence of such a DOS attack existing in the wild, by all means let me know.

Please open a new issue if you wish to continue discussion; I'm not sure why it was brought up in this PR.
